runtime/src/iree/io/stream.h

```diff
@@ -0,0 +1,228 @@
+
+// Maps a span of |length| bytes for writing. The stream offset is advanced by
+// |length| immediately and the caller must fill the returned |out_span| with
+// contents prior to closing the file.
+// Requires that IREE_IO_STREAM_MODE_WRITABLE and IREE_IO_STREAM_MODE_MAPPABLE
+// are set.
+IREE_API_EXPORT iree_status_t
+iree_io_stream_map_write(iree_io_stream_t* stream, iree_host_size_t length,
+                         iree_byte_span_t* out_span);
+
+// Copies |length| bytes from |source_stream| to |target_stream|.
+// Requires |source_stream| have IREE_IO_STREAM_MODE_READABLE and
+// |target_stream| have IREE_IO_STREAM_MODE_WRITABLE.
+IREE_API_EXPORT iree_status_t iree_io_stream_copy(
+    iree_io_stream_t* source_stream, iree_io_stream_t* target_stream,
+    iree_io_stream_pos_t length);
+
+//===----------------------------------------------------------------------===//
+// iree_io_stream_t implementation details
+//===----------------------------------------------------------------------===//
+
+typedef struct iree_io_stream_vtable_t {
+  void(IREE_API_PTR* destroy)(iree_io_stream_t* IREE_RESTRICT stream);
+
+  iree_io_stream_pos_t(IREE_API_PTR* offset)(iree_io_stream_t* stream);
+  iree_io_stream_pos_t(IREE_API_PTR* length)(iree_io_stream_t* stream);
+  iree_status_t(IREE_API_PTR* seek)(iree_io_stream_t* stream,
+                                    iree_io_stream_seek_mode_t seek_mode,
+                                    iree_io_stream_pos_t offset);
+  iree_status_t(IREE_API_PTR* read)(iree_io_stream_t* stream,
+                                    iree_host_size_t buffer_capacity,
+                                    void* buffer,
+                                    iree_host_size_t* out_buffer_length);
+  iree_status_t(IREE_API_PTR* write)(iree_io_stream_t* stream,
+                                     iree_host_size_t buffer_length,
+                                     const void* buffer);
+  iree_status_t(IREE_API_PTR* fill)(iree_io_stream_t* stream,
+                                    iree_io_stream_pos_t count,
+                                    const void* pattern,
+                                    iree_host_size_t pattern_length);
+  iree_status_t(IREE_API_PTR* map_read)(iree_io_stream_t* stream,
+                                        iree_host_size_t length,
+                                        iree_const_byte_span_t* out_span);
+  iree_status_t(IREE_API_PTR* map_write)(iree_io_stream_t* stream,
+                                         iree_host_size_t length,
+                                         iree_byte_span_t* out_span);
+} iree_io_stream_vtable_t;
+
+struct iree_io_stream_t {
+  iree_atomic_ref_count_t ref_count;
+  const iree_io_stream_vtable_t* vtable;
+  iree_io_stream_mode_t mode;
+};
+
+//===----------------------------------------------------------------------===//
+// iree_io_memory_stream_t
+//===----------------------------------------------------------------------===//
```

Comment:
Would it make sense to pull the implementations out into separate files? Maybe when `iree_io_buffer_stream_t`, `iree_io_file_stream_t`, etc. are implemented and the .c file grows > 1k LOC?

Reply:
memory and buffer stream are zero-dependency and DCEd if not used so there's no real need to put them in separate files (with separate deps etc) - file stream will exist in its own file as it has dependencies we don't want to pull in everywhere